modules/apps/asset/asset-taglib/src/main/resources/META-INF/resources/asset_categories_selector/AssetVocabularyCategoriesSelector.es.js

```diff
@@ -12,300 +12,276 @@
-/**
- * Wraps Clay's existing <code>MultiSelect</code> component that offers the user
- * a categories selection input.
- */
-class AssetVocabularyCategoriesSelector extends Component {
-	/**
-	 * @inheritDoc
-	 */
-	attached(...args) {
-		super.attached(...args);
-
-		this._dataSource = this._handleQuery.bind(this);
-	}
-
-	/**
-	 * Opens the tag selection dialog.
-	 *
-	 * @private
-	 */
-	_handleButtonClicked() {
-		const sub = (str, obj) => str.replace(/\{([^}]+)\}/g, (_, m) => obj[m]);
-
-		const uri = sub(decodeURIComponent(this.portletURL), {
-			selectedCategories: this.selectedItems
-				.map(item => item.value)
-				.join(),
-			singleSelect: this.singleSelect,
-			vocabularyIds: this.vocabularyIds.concat()
-		});
+	useEffect(() => {
+		ref.current = value;
+	}, [value]);
 
-		const itemSelectorDialog = new ItemSelectorDialog({
-			eventName: this.eventName,
-			title: Liferay.Language.get('select-categories'),
-			url: uri
-		});
+	return ref.current;
+}
 
-		itemSelectorDialog.open();
-		itemSelectorDialog.on('selectedItemChange', event => {
-			const selectedItems = event.selectedItem;
-
-			if (selectedItems) {
-				this.selectedItems = Object.keys(selectedItems).map(itemKey => {
-					return {
-						label: selectedItems[itemKey].value,
-						value: selectedItems[itemKey].categoryId
-					};
-				});
-			}
-		});
-	}
-
-	_handleInputFocus(event) {
-		this.emit('inputFocus', event);
-	}
-
-	/**
-	 * Converts the list of selected categories into a comma-separated serialized
-	 * version to be used as a fallback for old services and implementations.
-	 *
-	 * @private
-	 * @return {string} The serialized, comma-separated version of the selected items.
-	 */
-	_getCategoryIds() {
-		return this.selectedItems
-			.map(selectedItem => selectedItem.value)
-			.join();
-	}
-
-	/**
-	 * Shows the category error.
-	 *
-	 * @private
-	 */
-	_handleErrorAddinglabelItem(event) {
-		if (event.data.itemDoesNotExists) {
-			this._typedCategory = event.target.inputValue;
-			this._unexistingCategoryError = true;
+function AssetVocabulariesCategoriesSelector({
+	eventName,
+	id,
+	isValid = true,
+	groupIds = [],
+	inputName,
+	label,
+	onSelectedItemsChange = () => {},
+	portletURL,
+	required,
+	selectedItems = [],
+	singleSelect,
+	sourceItemsVocabularyIds = [],
+	useFallbackInput
+}) {
+	const [inputValue, setInputValue] = useState('');
+	const [invalidItems, setInvalidItems] = useState([]);
+
+	const {refetch, resource} = useResource({
+		fetchOptions: {
+			body: new URLSearchParams({
+				cmd: JSON.stringify({
+					'/assetcategory/search': {
+						'-obc': null,
+						end: 20,
+						groupIds,
+						name: `%${inputValue.toLowerCase()}%`,
+						start: 0,
+						vocabularyIds: sourceItemsVocabularyIds
+					}
+				}),
+				p_auth: Liferay.authToken
+			}),
+			credentials: 'include',
+			method: 'POST',
+			'x-csrf-token': Liferay.authToken
+		},
+		link: `${window.location.origin}${themeDisplay.getPathContext()}
+				/api/jsonws/invoke`
```

Comment:
Super scary way to build a link 😱 — seems there is some unwanted whitespace in here in any case, unless GitHub is deceiving me.

Reply:
Quick answer: GitHub is deceiving you and you know it ;)

Long answer: Yes, this is a scary way to build a link, but from what I've seen [here](https://github.com/liferay/clay/blob/6c97a7b0975e6a34c2901191bd60aaf1f0ea4ee5/packages/clay-data-provider/src/useResource.ts#L148-L153) it looks like we can either pass a string or a function for the `link` property. 

For the moment I won't be changing this as but we can keep on talking to see how it could be improved.